File: Baekjoon/12789 - 도키도키 간식드라미/12789.cpp
#include <iostream>
#include <stack>
#include <queue>
using namespace std;

int main() {
    cin.tie(NULL); cout.tie(NULL);
    ios::sync_with_stdio(false);

    int n, temp, now = 1;
    stack<int> st;
    queue<int> q;

    cin >> n;
    for(int i = 0; i < n; i++) {
        cin >> temp;

        q.push(temp);
    }

    while(q.size() || st.size()) {
        if(q.size() && now == q.front()) {
            now += 1;
            q.pop();
        }
        else if(st.size() && now == st.top()) {
            now += 1;
            st.pop();
        }
        else if(st.size() && !q.size() && now != st.top()) {
            break;
        }
        else {
            st.push(q.front());
            q.pop();
        }
    }

    if(q.size() || st.size()) cout << "Sad";
    else cout << "Nice";

    return 0;
}

File: solution/12xxx/12789 - 도키도키 간식드라미/12789.py
import sys
input = sys.stdin.readline

n = int(input())
ticket = list(map(int, input().split()))
stack = list()

result = "Nice"
now = 1

while len(ticket) + len(stack):
    if len(stack) != 0 and stack[-1] == now:
        stack.pop()
        now += 1
    elif len(ticket) != 0 and ticket[0] == now:
        ticket.pop(0)
        now += 1
    elif len(ticket) != 0 and ticket[0] != now:
        stack.append(ticket[0])
        ticket.pop(0)
    else:
        result = "Sad"
        break

print(result)